fix-Divide by zero result is wrong

WeiCheng14159 opened the pull request:
32-bit signed integer division in C is compiled to DIVW RISC-V instr. Given the divisor is zero, the quotient should be 0xff...f.
The result of IDIV instr (stored in EAX) isn't mapped to the **rd** register, and therefore, not seen by the guest program. 

Consider the test case in `test/test_programs/arithmetic_test/compiled_test_arithm.c` line 77
Change from 
```c=77
{
    //div-zero quotient should have all bits set
    init(++num, "DivZero");
    size_t n = 256;
    size_t m = 0;
    assert_equals(0xFFFFFFFFFFFFFFFF, (n / m), &failed_tests);
}
```
to 
```c=77
{
    //div-zero quotient should have all bits set
    init(++num, "DivZero");
    int n = 256;
    int m = 0;
    assert_equals(0xFFFFFFFFFFFFFFFF, (n / m), &failed_tests);
}
```
will result in a failed test and expose this problem

The fix provided is identical to the approach in **translate_DIV(...)** function
I guess you forget to add the same checking for DIVW (?)

WeiCheng14159 committed:
fix-Divide by zero result is wrong

32-bit signed integer division in C is compiled to DIVW RISC-V instr.
Given the divisor is zero, the quotient should be 0xff...f.
The result of IDIV instr (stored in EAX) isn't mapped to 'rd' register,
and therefore, not seen by guest program

flodt commented:
Yes, it seems as though we forgot to transfer the result to the listed rd. Thank you!
This looks good to me.

Simon04090 reviewed (changes requested):
Thank you for pointing out that mistake.
Since the DIVW instruction needs sign-extension, which is only done in the notDivByZero-Branch, it seems like we forgot to add the rd writeBack in the divByZero Branch.
Only a very minor thing to change that I would suggest is the following:
What this will essentially do now is (translated to normal Intel syntax)

```
<load registers>
TEST rs2, rs2
JZ divZero
CDQ
IDIV rs2
MOVSXD rd, EAX
JMP notDivZero
divZero:
MOV RAX, -1
notDivZero:
MOVSXD rd, EAX
```
So when rs2 is not zero, this does the movsxd twice, and when it is, it first moves -1 to RAX and then immediately moves it to rd again.
Instead I would say changing
`fe_enc64(&current, FE_MOV64ri, FE_AX, -1); ` 
to
`fe_enc64(&current, FE_MOV64ri, regDest, -1);`
should fix it as well and results in fewer emitted instructions.
Also the DIVUW instruction probably has the same mistake.

flodt commented:
Agreed.
`translate_DIVUW` does in fact contain the same mistake.
If I am not mistaken, the M-extension's instructions are also not yet covered by the parameterised unit tests, so that particular bug was able to slip through the cracks.

WeiCheng14159 committed:
feat-Add 32-bit divide by zero test case

Signed division in 32-bit will be compiled into DIVW
RISC-V instruction. This test case is created to test
DIVW specifically

WeiCheng14159 committed:
feat-Add 64-bit divide-by-zero test case

64-bit version of division will be compiled into DIV
and DIVU RISC-V instruction respectively. This test
case will test the correctness of these two instr on
boundary cases

WeiCheng14159 committed:
fix-Divide by zero result

32-bit signed integer division in C is compiled to DIVW RISC-V instr. Given the divisor is zero, the quotient should be 0xff...f. Which isn't mapped to the rd register, and therefore, not seen by the guest program

WeiCheng14159 committed:
fix-DIVUW divide by zero result

The divide-by-zero result (0xff...f) of DIVUW instruction should be mapped to the rd register

WeiCheng14159 commented:
@Simon04090 You are right indeed ! 
Thanks you both for the clarification :+1: 
I've changed the divide-by-zero handling in DIVW and DIVUW instructions as suggest by @Simon04090 
Also I've add some division test cases for 32/64 bit signed/unsigned division. Merge/Change it if your find them appropriate.

Simon04090 commented on test/test_programs/arithmetic_test/compiled_test_arithm.c:
```diff
@@ -75,12 +76,36 @@ int main(int argc, char **argv) {
         //div-zero quotient should have all bits set
-        init(++num, "DivZero");
+        init(++num, "DivZero (unsigned)");
         size_t n = 256;
```
This is also 64-bit unsigned since size_t := uint64_t (on 64 bit architectures).
So that test is double right now. I suggest changing here to either uint32_t or unsigned (int).

Simon04090 replied:
Decided to just quickly resolve this myself since I had the project loaded up and wanted to add the compiled binary anyways.

Simon04090 committed:
Fix wrong bitness

Simon04090 committed:
Fix wrong compare result and add binary.

DIVUW by zero has actually only the lower 32 bits set.
Also add the compiled binary.

WeiCheng14159 commented:
Looks good to me 👍
